modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/MetricNameValidator.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.telemetry.apm.internal;
+
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class MetricNameValidator {
+    private final Pattern ALLOWED_CHARACTERS = Pattern.compile("[a-z][a-z0-9_]*");
+    private final Set<String> ALLOWED_SUFFIXES = Set.of(
+        "size",
+        "total",
+        "count",
+        "usage",
+        "utilization",
+        "histogram",
+        "ratio",
+        "status" /*a workaround for enums */,
+        "time"
+    );
+
+    public static final int MAX_ELEMENT_LENGTH = 30;
+    public static final int MAX_NUMBER_OF_ELEMENTS = 10;
+
+    public void validate(String name) {
```

Comment:
Could this be static? The validation methods seem to be stateless, no need to have an instance?

Reply:
I tend to start with a regular class instead of a static util. Static utilities are creating tight coupling and prevent testing in isolation.
However I have moved the use of `MetricNameValidator` to `AbstractInstrument` and refactored all the test metric names to pass the rules so there probably will never be need to test instruments in isolation of the metric name validation.